service/header/store_heightsub.go

```diff
@@ -0,0 +1,84 @@
+package header
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+)
+
+// heightSub provides a minimalistic mechanism to wait till header for a height becomes available.
+type heightSub struct {
+	// height refers to the latest locally available header height
+	// that has been fully verified and inserted into the subjective chain
+	height       uint64 // atomic
+	heightReqsLk sync.Mutex
+	heightReqs   map[uint64][]chan *ExtendedHeader
+}
+
+// newHeightSub instantiates new heightSub.
+func newHeightSub() *heightSub {
+	return &heightSub{
+		heightReqs: make(map[uint64][]chan *ExtendedHeader),
+	}
+}
+
+// Height reports current height.
+func (hs *heightSub) Height() uint64 {
+	return atomic.LoadUint64(&hs.height)
+}
+
+// Sub subscribes for a header of a given height.
+// It can return both values as nil, which means a requested header was already provided
+// and caller should get it elsewhere.
+func (hs *heightSub) Sub(ctx context.Context, height uint64) (*ExtendedHeader, error) {
+	if hs.Height() >= height {
+		return nil, nil
+	}
+
+	hs.heightReqsLk.Lock()
+	if hs.Height() >= height {
+		// This is a rare case we have to account for.
+		// The lock above can park a goroutine long enough for hs.height to change for a requested height,
+		// leaving the request never fulfilled and the goroutine deadlocked.
+		return nil, nil
+	}
+	resp := make(chan *ExtendedHeader, 1)
+	hs.heightReqs[height] = append(hs.heightReqs[height], resp)
+	hs.heightReqsLk.Unlock()
+
+	select {
+	case resp := <-resp:
+		return resp, nil
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+// Pub processes all the outstanding subscriptions matching the given headers.
+// Pub is safe to be called from one goroutine.
+func (hs *heightSub) Pub(headers ...*ExtendedHeader) {
+	height := hs.Height()
+	from, to := uint64(headers[0].Height), uint64(headers[len(headers)-1].Height)
+	if height != 0 && height+1 != from {
+		log.Warnf("PLEASE REPORT THE BUG: headers given to the heightSub are in the wrong order")
```

Comment:
Shouldn't this be log.Errorf if this indicates a bug?

Reply:
Good point. Also, I think we should go even further and made it a Fatal error that terminates the node, otherwise storage might be unrecoverable

Reply:
Done.